File: app/src/main/java/com/tangem/data/Blockchain.java
package com.tangem.data;

import com.tangem.tangem_sdk.R;

/**
 * Created by dvol on 06.08.2017.
 */
public enum Blockchain {
    Unknown("", "", 1.0, R.drawable.ic_logo_unknown, "Unknown"),
    Bitcoin("BTC", "BTC", 100000000.0, R.drawable.ic_logo_bitcoin, "Bitcoin"),
    BitcoinTestNet("BTC/test", "BTC", 100000000.0, R.drawable.ic_logo_bitcoin_testnet, "Bitcoin Testnet"),
    BitcoinDual("BTC/dual", "BTC", 100000000.0, R.drawable.ic_logo_bitcoin, "Bitcoin"),
    Ethereum("ETH", "ETH", 1.0, R.drawable.ic_logo_ethereum, "Ethereum"),
    EthereumId("ETH/ID", "ETH", 1.0, R.drawable.ic_logo_ethereum, "Ethereum"),
    EthereumTestNet("ETH/test", "ETH", 1.0, R.drawable.ic_logo_ethereum_testnet, "Ethereum Testnet"),
    Token("Token", "ETH", 1.0, R.drawable.ic_logo_ethereum, "Ethereum"),
    NftToken("NftToken", "", 1.0, R.drawable.ic_logo_ethereum, "Ethereum"),
    BitcoinCash("BCH", "BCH", 100000000.0, R.drawable.ic_logo_bitcoin_cash, "Bitcoin Cash"),
    Litecoin("LTC", "LTC", 100000000.0, R.drawable.ic_logo_litecoin, "Litecoin"),
    Rootstock("RSK", "RBTC", 1.0, R.drawable.tangem2, "RSK"),
    RootstockToken("RskToken", "RBTC", 1.0, R.drawable.tangem2, "RSK"),
    Cardano("CARDANO", "ADA", 1000000.0, R.drawable.tangem2, "Cardano"),
    Ripple("XRP", "XRP", 1000000.0, R.drawable.ic_logo_xrp, "XRP"),
    Binance("BINANCE", "BNB", 100000000.0, R.drawable.ic_logo_binance, "Binance"),
    BinanceTestNet("BINANCE/test", "BNB", 100000000.0, R.drawable.ic_logo_binance, "Binance Testnet"),
    Matic("MATIC", "MTX", 1.0, R.drawable.tangem2, "Matic"),
    MaticTestNet("MATIC/test", "MTX", 1.0, R.drawable.tangem2, "Matic Testnet"),
    Stellar("XLM", "XLM", 10000000.0, R.drawable.ic_logo_stellar, "Stellar"),
    StellarTestNet("XLM/test", "XLM", 10000000.0, R.drawable.ic_logo_stellar, "Stellar Testnet"),
    StellarAsset("Asset", "XLM", 10000000.0, R.drawable.ic_logo_stellar, "Stellar"),
    StellarTag("XLM-Tag", "XLM", 1000000.0, R.drawable.ic_logo_stellar, "Stellar"),
    Eos("EOS", "EOS", 10000.0, R.drawable.tangem2, "EOS"),
    Ducatus("DUC", "DUC", 100000000.0, R.drawable.tangem2, "Ducatus"),
    Tezos("TEZOS", "XTZ", 10000000.0, R.drawable.ic_logo_tezos, "Tezos"),
    FlowDemo("FLOW/demo", "", 1.0, R.drawable.tangem2, "Flow demo");

    Blockchain(String ID, String currency, double multiplier, int imageResource, String officialName) {
        mID = ID;
        mCurrency = currency;
//        mMultiplier = multiplier;
        mImageResource = imageResource;
        mOfficialName = officialName;
    }

    private String mID, mOfficialName;
    //private double mMultiplier;
    private String mCurrency;
    private int mImageResource;

    public String getID() {
        return mID;
    }

    public String getOfficialName() {
        return mOfficialName;
    }

//    public double getMultiplier() {
//        return mMultiplier;
//    }

    public String getCurrency() {
        return mCurrency;
    }

    public static Blockchain fromId(String id) {
        for (Blockchain blockchain : values()) {
            if (blockchain.getID().equals(id)) return blockchain;
        }
        return Blockchain.Unknown;
    }

    public static Blockchain fromCurrency(String currency) {
        for (Blockchain blockchain : values()) {
            if (blockchain.getCurrency() == currency) return blockchain;
        }
        return null;
    }

    public static String[] getCurrencies() {
        String[] result = new String[values().length - 1];
        for (int i = 0; i < result.length - 1; i++) {
            result[i] = values()[i + 1].getCurrency();
        }
        return result;
    }

    private int getImageResource() {
        return mImageResource;
    }

    public int getLogoImageResource(String symbolName) {
        switch (this) {
            case Token:
                if (symbolName.equals("SEED"))
                    return R.drawable.ic_logo_seed;
                break;
            case RootstockToken:
                if (symbolName.equals("RIF"))
                    return R.drawable.ic_logo_rif;
        }

        return getImageResource();
    }

}

File: blockchain/src/main/java/com/tangem/blockchain/common/Blockchain.kt
package com.tangem.blockchain.common

import com.tangem.blockchain.blockchains.binance.BinanceAddressService
import com.tangem.blockchain.blockchains.bitcoin.BitcoinAddressService
import com.tangem.blockchain.blockchains.bitcoincash.BitcoinCashAddressService
import com.tangem.blockchain.blockchains.cardano.CardanoAddressService
import com.tangem.blockchain.blockchains.ethereum.EthereumAddressService
import com.tangem.blockchain.blockchains.stellar.StellarAddressService
import com.tangem.blockchain.blockchains.tezos.TezosAddressService
import com.tangem.blockchain.blockchains.xrp.XrpAddressService

enum class Blockchain(
        val id: String,
        val currency: String,
        val fullName: String
) {
    Unknown("", "", ""),
    Bitcoin("BTC", "BTC", "Bitcoin"),
    BitcoinTestnet("BTC/test", "BTCt", "Bitcoin Testnet"),
    BitcoinCash("BCH", "BCH", "Bitcoin Cash"),
    Litecoin("LTC", "LTC", "Litecoin"),
    Ducatus("DUC", "DUC", "Ducatus"),
    Ethereum("ETH", "ETH", "Ethereum"),
    RSK("RSK", "RBTC", "RSK"),
    Cardano("CARDANO", "ADA", "Cardano"),
    XRP("XRP", "XRP", "XRP Ledger"),
    Binance("BINANCE", "BNB", "Binance"),
    BinanceTestnet("BINANCE/test", "BNBt", "Binance"),
    Stellar("XLM", "XLM", "Stellar"),
    Tezos("TEZOS", "XTZ", "Tezos");

    fun decimals(): Int = when (this) {
        Bitcoin, BitcoinTestnet, BitcoinCash, Binance, BinanceTestnet, Litecoin, Ducatus -> 8
        Cardano, XRP, Tezos -> 6
        Ethereum, RSK -> 18
        Stellar -> 7
        Unknown -> 0
    }

    fun pendingTransactionsTimeout(): Int = when (this) {
        else -> 0
    }

    fun makeAddress(walletPublicKey: ByteArray): String = getAddressService().makeAddress(walletPublicKey)

    fun validateAddress(address: String): Boolean = getAddressService().validate(address)

    private fun getAddressService(): AddressService = when (this) {
        Bitcoin, BitcoinTestnet, Litecoin, Ducatus -> BitcoinAddressService(this)
        BitcoinCash -> BitcoinCashAddressService()
        Ethereum, RSK -> EthereumAddressService()
        Cardano -> CardanoAddressService()
        XRP -> XrpAddressService()
        Binance -> BinanceAddressService()
        BinanceTestnet -> BinanceAddressService(true)
        Stellar -> StellarAddressService()
        Tezos -> TezosAddressService()
        Unknown -> throw Exception("unsupported blockchain")
    }

    fun getShareUri(address: String): String = when (this) {
        Bitcoin -> "bitcoin:$address"
        Ethereum -> "ethereum:$address"
        XRP -> "ripple:$address"
        else -> address
    }

    fun getExploreUrl(address: String, token: Token? = null): String = when (this) {
        Binance -> "https://explorer.binance.org/address/$address"
        BinanceTestnet -> "https://testnet-explorer.binance.org/address/$address"
        Bitcoin -> "https://blockchain.info/address/$address"
        BitcoinTestnet -> "https://live.blockcypher.com/btc-testnet/address/$address"
        BitcoinCash -> "https://blockchair.com/bitcoin-cash/address/$address"
        Litecoin -> "https://live.blockcypher.com/ltc/address/$address"
        Ducatus -> "https://insight.ducatus.io/#/DUC/mainnet/address/$address"
        Cardano -> "https://cardanoexplorer.com/address/$address"
        Ethereum -> if (token == null) {
            "https://etherscan.io/address/$address"
        } else {
            "https://etherscan.io/token/${token.contractAddress}?a=$address"
        }
        RSK -> {
            var url = "https://explorer.rsk.co/address/$address"
            if (token != null) {
                url += "?__tab=tokens"
            }
            url
        }
        Stellar -> "https://stellar.expert/explorer/public/account/$address"
        XRP -> "https://xrpscan.com/account/$address"
        Tezos -> "https://tezblock.io/account/$address"
        Unknown -> throw Exception("unsupported blockchain")
    }

    companion object {
        private val values = values()
        fun fromId(id: String): Blockchain = values.find { it.id == id } ?: Unknown
        fun fromName(name: String): Blockchain = values.find { it.name == name } ?: Unknown
        fun fromCurrency(currency: String): Blockchain = values.find { it.currency == currency }
                ?: Unknown
    }
}
